uaclient/tests/test_cli_attach.py

```diff
@@ -86,23 +88,26 @@ def test_happy_path_without_arg(self, action_status, contract_client,
         expected_macaroon = bound_macaroon.decode('utf-8')
         assert expected_macaroon == cfg._cache_contents['bound-macaroon']
 
-    @mock.patch('uaclient.cli.sso.discharge_root_macaroon')
-    @mock.patch('uaclient.cli.contract.UAContractClient')
-    @mock.patch('uaclient.cli.action_status')
-    def test_happy_path_with_arg(self, action_status, contract_client,
-                                 discharge_root_macaroon):
-        """A mock-heavy test for the happy path without an argument"""
+    @mock.patch(M_PATH + 'sso.discharge_root_macaroon')
+    @mock.patch(
+        M_PATH + 'contract.UAContractClient.request_contract_machine_attach')
+    @mock.patch(M_PATH + 'action_status')
+    def test_happy_path_with_token_arg(self, action_status,
+                                       contract_machine_attach,
+                                       discharge_root_macaroon):
+        """A mock-heavy test for the happy path with the contract token arg"""
         # TODO: Improve this test with less general mocking and more
         # post-conditions
-        token = 'my-token'
+        token = 'contract-token'
         args = mock.MagicMock(token=token)
         cfg = TestConfig.with_account()
 
         ret = action_attach(args, cfg)
 
         assert 0 == ret
         assert 1 == action_status.call_count
-        assert token == cfg._cache_contents['bound-macaroon']
+        contract_machine_attach.called_once_with(contract_token=token)
```

Comment:
The assertion methods on mocks are best avoided, because typos in the names of the methods are silently swallowed:

```
>>> my_mock = mock.Mock()
>>> my_mock()
<Mock name='my_mock()' id='139900840650568'>
>>> my_mock.assert_not_called()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3/dist-packages/mock/mock.py", line 897, in assert_not_called
    raise AssertionError(msg)
AssertionError: Expected 'my_mock' to not have been called. Called 1 times.
>>> my_mock.asert_not_called()
<Mock name='mock.asert_not_called()' id='139900841774720'>
```

This is such a problem, in fact, that mock tries to catch some of these cases, but it obviously won't be perfect: https://github.com/python/cpython/blob/master/Lib/unittest/mock.py#L595

For these two, I would propose:

```py
assert [mock.call(contract_token=token)] == contract_machine_attach.call_args_list
assert 0 == discharge_root_macaroon.call_count
```

In both of these cases, if you typo the attribute on the mock, the assertion will fail.

Reply:
+1 changed unit tests to avoid error-prone typos on mock asserts. Thanks for this +1

Reply:
Perhaps GH is slow to update, but it looks like we're still using `called_once_with`?

Reply:
still not resolved

Reply:
thanks . just pushed a fix for this unit test